design/design.md

```diff
@@ -0,0 +1,251 @@
+serializing a certificate if the number of allowed repetitions $r = 1$.
+The simple lottery scheme, on the other hand, has large proof sizes but offers
+very fast proving time which might be important for some applications.
+Additionally, in the decentralized setting, it has the smallest communication
+complexity $\mu$, so it's likely to be the top choice for applications where
+the certificate size is not important.
+
+We will implement schemes that seem redundant (e.g., unweighted schemes
+since the user could just use a weighted variant with each weight equal 1,
+or centralized schemes since the user could just use a decentralized variant
+directly) for two reasons:
+1. to provide a simpler interface for the user;
+2. because they might be more efficient now or when optimizations are
+implemented in the future; for instance, unweighted variants can be concretely
+more efficient due to no lottery and less complexity, and in the decentralized
+setting, an unweighted version could be more efficient due to not needing to
+sample the binomial distribution.
+
+We should be able to reuse most of the logic between different schemes anyway,
+so the overhead shouldn't be large.
+
+## Choice of hash function
+The "random" functions $H_0$, $H_1$, $H_2$ and others in ALBA constructions need to be
+implemented using some concrete function, either a cryptographic hash function
+modeled as a random oracle or a
+[pseudorandom function](https://en.wikipedia.org/wiki/Pseudorandom_function_family).
+There are numerous options available for one or both models:
+SHA, BLAKE, AES, ChaCha, etc.
+
+In our implementation, we will let the user provide his own function.
+Hash functions evolve, old ones get broken (SHA-1), and so our ALBA
+implementation shouldn't get stuck with an inefficient / broken hash function.
+Moreover, the user might want to instantiate ALBA with a PRF (e.g. AES) instead
+of a hash function modeled as a random oracle for better theoretical guarantees.
+With Rust's advanced trait system, making the hash function generic should be
+easy, and the runtime code performance should be fast due to compile time
+function resolution.
+
+We will provide an example implementing a hash function with the right interface
+for ALBA using, for example, BLAKE underneath, but with a warning that
+it is provided for demonstration / prototyping only and should not be used in
+production.
+
+One reason is that when ALBA is used alongside other sub-protocols that also
+need hashing, it is the user's responsibility to implement
+[domain separation](https://en.wikipedia.org/wiki/Domain_separation)
+to provide the different
+sub-protocols with independent hash functions.
+Otherwise, the underlying assumptions needed for security might break and the
+larger protocol can become insecure
+(see [here](https://link.springer.com/chapter/10.1007/978-3-030-45724-2_1)).
+
+Additionally, different hash functions (e.g., SHA and BLAKE) shouldn't mix and
```

Comment:
This sentence is not very clear to a user.

Reply:
done